lib/lrama/option_parser.rb

```diff
@@ -0,0 +1,124 @@
+require 'optparse'
+
+module Lrama
+  # Handle option parsing for the command line interface.
+  class OptionParser
+    def initialize
+      @options = Options.new
+      @trace = []
+      @report = []
+    end
+
+    def parse(argv)
+      parse_by_option_parser(argv)
+
+      @options.trace_opts = validate_trace(@trace)
+      @options.report_opts = validate_report(@report)
+      @options.grammar_file = argv.shift
+
+      if !@options.grammar_file
+        abort "File should be specified\n"
+      end
+
+      if @options.grammar_file == '-'
+        @options.grammar_file = argv.shift or abort "File name for STDIN should be specified\n"
+      else
+        @options.y = File.open(@options.grammar_file, 'r')
+      end
+
+      if !@report.empty? && @options.report_file.nil? && @options.grammar_file
+        @options.report_file = File.dirname(@options.grammar_file) + "/" + File.basename(@options.grammar_file, ".*") + ".output"
+      end
+
+      if !@options.header_file && @options.header
+        case
+        when @options.outfile
+          @options.header_file = File.dirname(@options.outfile) + "/" + File.basename(@options.outfile, ".*") + ".h"
+        when @options.grammar_file
+          @options.header_file = File.dirname(@options.grammar_file) + "/" + File.basename(@options.grammar_file, ".*") + ".h"
+        end
+      end
+
+      @options
+    end
+
+    private
+
+    def parse_by_option_parser(argv)
+      ::OptionParser.new do |o|
+        o.banner = <<~BANNER
+          Lrama is LALR (1) parser generator written by Ruby.
+
+          Usage: #{$0} [options] FILE
```

Comment:
`$0` sometimes verbosely print full path of program file like your example in PR description. I think it's better to print only base name ("lrama").

In this case, the name ("lrama") is static then it's ok to rewrite this banner like

```ruby
Usage: lrama [options] FILE
```

Default value of `program_name` is `File.basename($0, '.*')`. https://github.com/ruby/optparse/blob/a291ef5c26445d423bf0223b0cb64340df18f0cb/lib/optparse.rb#L1248

Or another approach is using `#banner=` and `#banner` like

```ruby
      opt = OptionParser.new do |parser|
        parser.banner += " FILE"
      end
```

https://github.com/ruby/optparse/blob/a291ef5c26445d423bf0223b0cb64340df18f0cb/lib/optparse.rb#L104 is an example.

Reply:
Indeed it is. I will change it to use the static "lrama" string.